W-14222920-ping-cloud-duke (#214)
mulesoft/docs-code-builder: modules/ROOT/pages/int-trigger-flows.adoc

include::reuse::partial$beta-banner.adoc[tag="anypoint-code-builder"]

> When an integration or implementation that you are developing builds successfully within the Anypoint Code Builder IDE, the application deploys locally, within your development environment, whether on your desktop or the cloud-based host for your applications.

Reviewer: ```suggestion
When your integration or implementation builds successfully within the Anypoint Code Builder IDE, the application deploys to your local development environment, whether you are using the desktop or cloud-based host for your applications.
```

It is common to configure the HTTP Listener from  xref:http-connector::index.adoc[Anypoint Connector for HTTP] (HTTP Connector) to trigger the execution of a flow in a deployed application. You can trigger the flow locally from the built-in terminal in Anypoint Code Builder or from a VS Code extension like Thunder Cloud.

[[before-you-begin]]
== Before You Begin

* Complete xref:start-acb.adoc[] for your IDE.
* Make sure an integration or implementation project with an HTTP Listener (`<http:listener/>`) is available from the cloud or desktop IDE.
+
To create a project, see xref:int-create-integrations.adoc[] or xref:imp-implementing-apis.adoc[], or follow the procedures in the tutorial xref:tut-af-amflights-overview.adoc[].

[[trigger-urls]]
== URLs for Triggering a Flow from an HTTP Listener

For integrations and implementations that you build and deploy within your IDE during the development process, the URL to use for triggering an HTTP Listener differs depending on whether the listener is within the configuration XML for a scaffolded interface that is generated from an API specification or within a configuration XML file that does not implement a scaffolded interface.

> Typically, apps that use an HTTP Listener configure the host to `0.0.0.0` and the port to `8081` or a similar port, such as `8082`, if `8081` is currently in use by another application.

Reviewer: ```suggestion
Typically, for apps that use an HTTP Listener you configure the host to `0.0.0.0` and the port to `8081`. If port `8081` is in use by another application, you can use a similar port, such as `8082`.
```

[[trigger-url-no-interface]]
* Trigger a flow within a configuration XML that _does not_ contain a scaffolded interface by using a URL similar to this one from your IDE's terminal: 
+
----
http://localhost:8081/mypath
----
+
The port and host are defined in the `<http:listener-connection/>` element, and the endpoint is defined by the `path` attribute in the `<http:listener/>` element:
+
[source,xml]
----
<http:listener-config name="inbound-request" 
      doc:name="Listener Config" doc:id="b5c62f">
  <http:listener-connection host="0.0.0.0" port="8081"/>
</http:listener-config>
<flow name="myFlow">
    <http:listener path="mypath" config-ref="inbound-request" doc:name="HTTP /flights" />
    ...
</flow>
----

> [[trigger-url-interface]]

Reviewer: Is there a missing header?

Author: Theres no header for this. I link to this bullet from elsewhere later.

> * Trigger a flow from listener within a scaffolded interface by using a URL similar to this one from your IDE's terminal: 

Reviewer: ```suggestion
* Trigger a flow from the listener within a scaffolded interface by using a URL similar to this one from your IDE's terminal: 
```

<flow name="american-flight-api-main">
    <http:listener config-ref="inbound-request" path="/api/*"> <!--1-->
        <http:response statusCode="#[vars.httpStatus default 200]">
            <http:headers>#[vars.outboundHeaders default {}]</http:headers>
        </http:response>
        <http:error-response statusCode="#[vars.httpStatus default 500]">
            <http:body>#[payload]</http:body>
            <http:headers>#[vars.outboundHeaders default {}]</http:headers>
        </http:error-response>
    </http:listener>
    <apikit:router config-ref="american-flight-api-config" /> <!--2-->
    ...
</flow>
----
[calloutlist]
.. Within the listener (`<http:listener/>`), the attribute `path="/api/*"` is set automatically during the scaffolding process. 
.. The router (`<apikit:router/>`) is located below the listener in the scaffolded interface within the configuration XML. 

[[trigger-flow]]
== Trigger a Flow

Use the terminal in Anypoint Code Builder to trigger a flow through an HTTP Listener in an integration or implementation project that is deployed locally, within your IDE. This procedure is for Mule apps within the desktop or cloud IDE. 

> TIP: The examples in the procedure show how to trigger the flow with a `curl` command from the IDE's terminal. You can also use Thunder Cloud, a third-party extension to VS Code, to trigger a flow with the same URLs. For security, external REST clients, such as Advanced REST Client and Postman, are not allowed to send requests to projects in the cloud IDE.

Reviewer: ```suggestion
TIP: The examples in this procedure show how to trigger the flow with a `curl` command from the IDE's terminal. You can also use Thunder Cloud, a third-party extension for VS Code, to trigger a flow with the same URLs. For security reasons, external REST clients, such as Advanced REST Client and Postman, are not allowed to send requests to projects in the cloud IDE.
```

> Before attempting to trigger a flow, review <<trigger-urls>>, and be sure to meet the prerequisites in <<before-you-begin>>.

Reviewer: ```suggestion
Before attempting to trigger a flow, review <<trigger-urls>>, and verify that you meet the prerequisites in <<before-you-begin>>.
```

//first generic steps
include::anypoint-code-builder::partial$acb-reusable-steps.adoc[tags="check-connection-http"]
include::anypoint-code-builder::partial$acb-reusable-steps.adoc[tags="open-run-debug"]
include::anypoint-code-builder::partial$acb-reusable-steps.adoc[tags="deploy-locally"]
//open terminal
include::anypoint-code-builder::partial$acb-reusable-steps.adoc[tag="open-terminal"]
. In the IDE's terminal, click the name of the shell to use, such as PowerShell (for Windows), Bash (`bash`), or Z shell (`zsh` for Mac).
+
Alternatively, click the *+* icon in the terminal to open a new command prompt.
. At the command prompt, provide a `curl` command to trigger the flow, for example:

> 

Reviewer: Is the example missing? Or is this a duplicated line?

Author: fixed, thx for heads up

> . At the command prompt, provide a `curl` command to trigger the flow, for example:

Author: ```suggestion
. At the command prompt, provide a `curl` command to trigger the flow.
```

> * To trigger a flow from an HTTP listener that does not reside within an interface, use `curl` with a URL similar to this one:

Reviewer: ```suggestion
* To trigger a flow from an HTTP Listener that doesn't reside within an interface, use `curl` with a URL similar to this one:
```
